[jedidiahC] iP

jedidiahC opened the pull request:
# iPman
> "Be like water, my friend" - iPman's [mentor](https://en.wikipedia.org/wiki/Bruce_Lee)

iPman is the **one inch punch** to
- _Revolutionize_ your productivity 🔥
- _Cure_ cancer 🧘
- _Sleep_ peacefully at night 🌙

**Features**
- [x] Track tasks
- [ ] Cure cancer
- [ ] Cure insomnia

**To accept iPman as your ~~lord and saviour~~ mentor**:
1. Download it.
2. Run it.
3. Feed it.

**If iPman rebels against his creator(s), would you kindly remove this block of code:**
```java
public class Main {
    public static void main(String[] args) {
        new iPman().live()
    }
}
```

Commit:
Add Gradle support

jedidiahC committed:
Implemented Level 1 - Greet, Echo, Exit

jedidiahC committed:
Implemented Add and List.

jedidiahC committed:
Implemented Level 3 - marking and unmarking of tasks.

jedidiahC committed:
Implemented Level 4 - ToDo, Event, Deadline.

jedidiahC committed:
Added test cases for Level 4 - List, Todo, Deadline, Event, Mark, Unmark.

jedidiahC committed:
Implement error handling for Todo, Event, Deadline and refactored query handling

Refactoring query processing logic ensures separation of concerns in regards to processing the user query for different purposes.

It makes it easier to introduce different kinds of query in the future as well.

jedidiahC committed:
Implement Level 6 - Delete

jedidiahC committed:
Change query processing to use enums

jedidiahC committed:
Add task saving and loading

jedidiahC committed:
Change deadlines and events to use DateTime instead String

jedidiahC committed:
Merge branch 'branch-Level-7' into master

jedidiahC committed:
Merge branch 'branch-Level-8' into master

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/DeadlineQueryHandler.java
#	src/main/java/Event.java
#	src/main/java/TaskTracker.java

jedidiahC committed:
Refactor code to be more OOP

Code was violating SRP.

Let's refactor all saving / loading of tasks to TaskStorage.
Let's refactor all interactions with the user to Ui.

jedidiahC committed:
Divide classes into packages

jedidiahC committed:
Merge branch 'add-gradle-support' into master

jedidiahC committed:
Configure gradle to build and run Duke

jedidiahC committed:
Add JUnit tests for TaskTracker methods addDeadline, markTask and deleteTask

jedidiahC committed:
Fix delete, mark, unmark task success response

jedidiahC committed:
Add JavaDoc comments for some non-private classes/methods

jedidiahC committed:
Tweak the code to comply with the Java coding standard

jedidiahC committed:
Add way for users to find task by searching for a keyword

jedidiahC committed:
Merge branch 'branch-Level-9' into master

jedidiahC committed:
Merge branch 'branch-A-CodingStandard' into master

# Conflicts:
#	src/main/java/duke/Bot.java
#	src/main/java/duke/query/QueryTypeUtil.java

jedidiahC committed:
Merge branch 'branch-A-JavaDoc' into master

# Conflicts:
#	src/main/java/duke/query/EventQueryHandler.java
#	src/main/java/duke/query/IQueryHandler.java
#	src/main/java/duke/task/TaskTracker.java

jedidiahC committed:
Fix JavaDoc comments to follow coding convention

jedidiahC committed:
Add checkstyle config files

praveenkrishna0512 commented on src/main/java/duke/Bot.java:
```diff
@@ -0,0 +1,97 @@
+ * The Bot class represents a bot that can process and respond to user queries.
+ */
+public class Bot {
+    private static final boolean SHOULD_LOAD_TASK_SAVE = true;
```
May I clarify how this variable is used? Perhaps this variable could be named to convey this information to any reader :)

praveenkrishna0512 commented on src/main/java/duke/Bot.java:
```diff
@@ -0,0 +1,97 @@
+        StringTokenizer query = new StringTokenizer(input);
+        QueryType queryType = QueryType.UNKNOWN;
+
+        if (query.hasMoreTokens()) {
```
I like the naming for the `hasMoreTokens()` function! It conveys the purpose of the function well.

praveenkrishna0512 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,41 @@
+     * Begins and runs the bot's loop.
+     */
+    public void run() {
+        boolean running = true;
```
This is a minor issue but is there a reason why this variable is named `running`? Perhaps `isRunning` could be considered in accordance with the coding standard.

jedidiahC replied:
Great suggestion!

praveenkrishna0512 commented on src/main/java/duke/query/DeadlineQueryHandler.java:
```diff
@@ -0,0 +1,56 @@
+     */
+    @Override
+    public String processQuery(String query) throws DukeException {
+        String[] parsed = QueryParser.parseQuery(query, new String[]{"/by"});
```
May I ask why this variable was named `parsed`? Perhaps a name could be chosen to convey the message that is contains the arguments from the parsed command!

jedidiahC replied:
Yup, this is a good point. I will probably refactor the results of this function into its own class to better convey the intent and context behind the results.

praveenkrishna0512 commented on src/main/java/duke/query/EventQueryHandler.java:
```diff
@@ -0,0 +1,62 @@
+    public String processQuery(String query) throws DukeException {
+        String[] parsed = QueryParser.parseQuery(query, new String[]{"/from", "/to"});
+        String desc = parsed[1];
+        if (desc == null || desc.isBlank()) {
```
I like the naming of the `isBlank()` function, to convey the intention of the function!

praveenkrishna0512 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,69 @@
+        this.isDone = isDone;
+    }
+
+    protected void mark() {
```
May I clarify why the function is named `mark()`? Perhaps `markDone()` could be considered, to better convey the intent that this function marks the task as done.

praveenkrishna0512 reviewed (commented):
Overall, I believe that coding standards were adhered to! I like the way you have structured your code as well :) I left a few comments regarding minor adjustments that could be made to convey the intentions of functions clearly. Good job!

jedidiahC committed:
Add basic graphical user interface

jedidiahC committed:
Update styling for DialogBox

jedidiahC committed:
Merge branch 'branch-Level-10' into master

jedidiahC committed:
Add assertions for marking and unmarking tasks

jedidiahC committed:
Clean up several classes to improve code quality

After adding GUI, some classes are dead code. Also, the string array
which represents a parsed query is badly designed and does not
show intent well.

Extracting the parse result into its own object will make it easier
to access the various components of a query.

By refactoring query data to a Query object, extracting data from a
query is now less prone to error.

jedidiahC committed:
Merge pull request #1 from jedidiahC/branch-A-Assertions

Add assertions for marking and unmarking tasks

jedidiahC committed:
Merge branch 'master' into branch-A-CodeQuality

jedidiahC committed:
Merge pull request #2 from jedidiahC/branch-A-CodeQuality

Clean up several classes to improve code quality

jedidiahC committed:
Remove unnecessary println() from DeadlineQueryHandler

jedidiahC committed:
Add scalable dialog box

jedidiahC committed:
Move all task related query handling to new package duke.query.task

All query handling are in duke.query.

Moving all module specific query handling to its own package will
unclutter the query package.

jedidiahC committed:
Add loan tracking system

Bot should provide a service that supports the management of loan
records.

jedidiahC committed:
Merge branch 'branch-D-Loans' into master

jedidiahC committed:
Extract query processing logic from bot to QueryProcessor class

jedidiahC committed:
Remove unnecessary imports from DukeGUI and change interface IQueryHandler to abstract class QueryHandler

jedidiahC committed:
Extract duplicate query param extraction to QueryHandler superclass

jedidiahC committed:
Overhaul GUI look and feel

jedidiahC committed:
Add border radius styling to dialogue box

jedidiahC committed:
Add resizing of window

jedidiahC committed:
Tweak dialogue box look

jedidiahC committed:
Add min width and height of window

jedidiahC committed:
Tweak look and feel and fix bye query

jedidiahC committed:
Reorganise classes with new packages and add greeting messages

jedidiahC committed:
Add loan record query

jedidiahC committed:
Add flavour text for bot

jedidiahC committed:
Remove query type enum and refactor command matching logic

jedidiahC committed:
Add app icon and improve several responses

jedidiahC committed:
Add persistent data for loans

jedidiahC committed:
Add loan-summary query

jedidiahC committed:
Add a help system

jedidiahC committed:
Improve formatting for some responses

jedidiahC committed:
Add representative screenshot

jedidiahC committed:
Fix code convention violations and add javadoc comments

jedidiahC committed:
Update README.md

jedidiahC committed:
Create _config.yml

jedidiahC committed:
Update README.md

jedidiahC committed:
Add additional JUnit tests

jedidiahC committed:
Fix loan storage bug

jedidiahC committed:
Improve error handling for query handlers

jedidiahC committed:
Update README.md
